internal/provider/resource_ssh_key.go

```diff
@@ -5,183 +5,309 @@ import (
 
-	// create again
-	err = client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
-		ModelName: modelName,
-		ModelUUID: modelUUID,
-		Payload:   d.Get("payload").(string),
+	// Get the model name from the plan because it might have changed
+	newModelName := state.ModelName.ValueString()
+	newModelUUID, err := s.client.Models.ResolveModelUUID(newModelName)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
+	}
+
+	// Create a new key
+	err = s.client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
+		ModelName: newModelName,
+		ModelUUID: newModelUUID,
+		Payload:   plan.Payload.ValueString(),
 	})
 	if err != nil {
-		diags = diag.FromErr(err)
-		return diags
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create ssh key for updating, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("ssh key created : %q", plan.ID.ValueString()))
 
-	return diags
+	// Set the plan onto the Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
-func sshKeyDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+// Delete is called when the provider must delete the resource. Config
+// values may be read from the DeleteRequest.
+//
+// If execution completes without error, the framework will automatically
+// call DeleteResponse.State.RemoveResource(), so it can be omitted
+// from provider logic.
+//
+// Juju refers to deletion as "destroy" so we call the Destroy function of our client here rather than delete
+// This function remains named Delete for parity across the provider and to stick within terraform naming conventions
+func (s *sshKeyResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "ssh_key", "delete")
+		return
+	}
 
-	var diags diag.Diagnostics
+	var data sshKeyResourceModel
 
-	modelName := d.Get("model").(string)
-	modelUUID, err := client.Models.GetModelByName(modelName)
+	// Read Terraform configuration from the request into the data model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	modelName := data.ModelName.ValueString()
+	modelUUID, err := s.client.Models.ResolveModelUUID(modelName)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
 	}
 
-	user := utils.GetUserFromSSHKey(d.Get("payload").(string))
+	payload := data.Payload.ValueString()
+	user := utils.GetUserFromSSHKey(payload)
 	if user == "" {
-		return diag.Errorf("malformed SSH key, user not found")
+		resp.Diagnostics.AddError("Client Error", "malformed SSH key, user not found")
+		return
 	}
```

Comment:
Well I suppose if the user is removed externally before delete --or update-- (can a user be changed in juju like a rename?) then you'd get a user not found error from Juju anyways, so yeah why not

Reply:
User in the sshkey has nothing to do with juju users. It's the email address at the end. Therefore it can't change unless the data.Payload has. Then bigger changes would happen.

Reply:
oh duh, that makes sense :) yeah I changed it anyway